lib/utils/logDeprecation.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+const chalk = require('chalk');
+const memoizee = require('memoizee');
+
+const loggedDeprecations = {};
+
+function consoleLog(...args) {
+  console.log(...args); // eslint-disable-line no-console
+}
+
+function extractCodes(codesStr) {
+  if (!codesStr) {
+    return [];
+  }
+  return codesStr.split(',').reduce((acc, code) => {
+    acc[code] = true;
+    return acc;
+  }, {});
+}
+
+const prepareCodes = memoizee((serviceConfigCodesStr, envCodesStr) => {
+  const serviceConfigCodes = extractCodes(serviceConfigCodesStr);
+  const envCodes = extractCodes(envCodesStr);
+  const disabledCodes = Object.assign({}, envCodes, serviceConfigCodes);
+  return disabledCodes;
+});
+
+function writeDeprecation(message, docURL) {
+  consoleLog(`${chalk.red(message)}\n${chalk.dim(`More Info: ${docURL}`)}`);
+}
+
+function logDeprecation(code, message, docURL, serviceConfig) {
+  const disabledCodes = prepareCodes(
+    serviceConfig && serviceConfig.disabledCodes,
+    process.env.SLS_DEPRECATION_DISABLE
+  );
+  if (loggedDeprecations[code] || disabledCodes[code] || disabledCodes['*']) {
```

Comment:
I'd check for `loggedDeprecations` and `deprecatedByEnv` before we do any resolution from `serviceConfg`

Reply:
`extractCodes` function will return an empty set if any falsey value was passed. So it will always check for codes from all of them and `serviceConfig` codes will have a value only when passed

Reply:
My point was that `serviceConfig` should not be processed (resolved) if we know that deprecation should not be logged from other sources

Reply:
I wanted the if condition to be as small and as readable as possible

Reply:
Sure, I totally agree. What I proposed I believe doesn't change that.

What I meant is something as:

```javascript
if (loggedDeprecations.has(code) || deprecatedByEnv.has(code) || resolveDeprecatedByService(serviceConfig).has(code)) {
  return;
}
```

Reply:
Done
I separated it in a different if condition though